### PR TITLE
[WFCORE-6375][WFCORE-6434] Do not synchronize stopped servers when a HC reconnects and allow reload/restart required when starting

### PR DESCRIPTION
**Pull request by yersan**
Jira issue: https://issues.redhat.com/browse/WFCORE-6375
Jira issue: https://issues.redhat.com/browse/WFCORE-6434

### CODE REVIEW

File: host-controller/src/main/java/org/jboss/as/domain/controller/operations/SyncServerStateOperationHandler.java
Comment (bstansberry) [inline host-controller/src/main/java/org/jboss/as/domain/controller/operations/SyncServerStateOperationHandler.java:131]: In our June 1 chat we talked about logging a WARN here, not throwing an exception. I think this is fine though; if this is null it's a ??? situation and we should probably fail.

File: controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java:136]: Changing the value of newStamp instead of using a new variable makes me nervous, as this executes in a loop, so the reassignment can happen more than once. It might be fine, e.g. because of the synchronization on service, but it feels fragile / easily broken by a future maintainer.

**Review by bstansberry — CHANGES_REQUESTED** [review]
Hi @yersan. Good job; I like the ControlledProcessState approach. I have a couple minor change requests but once you've handled those feel free to merge.

**Comment by yersan** [comment]
@bstansberry I think the transition on starting still required a bit more thinking since the states of the newly added flags are not synched in the same way as the reference. The state of these flags is part of the state kept in the reference. I moved it to a draft until getting more confident with this

File: host-controller/src/main/java/org/jboss/as/domain/controller/operations/SyncServerStateOperationHandler.java
Comment (bstansberry) [reply host-controller/src/main/java/org/jboss/as/domain/controller/operations/SyncServerStateOperationHandler.java:131]: 👍

File: controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java:124]: I just noticed a race here. I think this should be:

int currentStamp = stamp.getAndIncrement();
int newStamp = currentStamp + 1; // the value from the increment we just did is "ours"

**Review by bstansberry — CHANGES_REQUESTED** [review]

File: controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java
Comment (yersan) [reply controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java:124]: If we increment it, we won't be able to revert from a restart/reload required. The issue is the stamp saved in the step that executed the setRestartRequired()/setReloadRequired() does not change, because it is an artificial state derived from a flag variable

Comment (yersan) [reply controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java:124]: If you want to see it in a different way, the `stamp.getAndIncrement()` you are requesting to do here, was already done when setRestartRequired()/setReloadRequired() was called: [ControlledProcessState.java#L192](https://github.com/wildfly/wildfly-core/blob/757e3012d9a14e3090fc129c22a82b15514d3e58/controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java#L192)

**Comment by yersan** [comment]
Hi @bstansberry , I'm fine with the current status, could you take a look again? thanks!

**Review by bstansberry — CHANGES_REQUESTED** [review]
@yersan Below I point out 3 places where I'd like changes. They all revolve around L123 in your current branch, which has the effect of using 'stamp' as part of the state machine and not solely as a unique value generator. That works but it's harder to understand and thus is fragile.

To review this I checked out your branch to do my reviewing in my IDE, and while doing that I edited some to help me understand. In the end doing that led to a proposed change that addresses the issues I mention below, plus adds comments and makes some other misc adjustments. I've sent you a PR with that. It passes the controller module tests and testsuite/domain. 

https://github.com/yersan/wildfly-core/pull/7

File: controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java
Comment (yersan) [reply controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java:136]: > This use of currentStamp treats 'stamp' as part of the state machine.

Yes, thanks @bstansberry , I understand where my approach was failing, due to the thread scheduling, my approach could make it impossible to revert the restart/reload required.

Your suggestion prevents the stamp stored in the step could be different than the one expected to revert.

**Review by bstansberry — APPROVED** [review]
Thanks @yersan LGTM.

I had a thought this morning for further improvement in the future. Create a new InternalState enum that encapsulates State, reloadRequiredOnStarting and restartRequiredOnStarting. Then the 'state' var stores InternalState, not State. Then we just do state transitions, instead of manipulating multiple fields.  Besides probably being cleaner looking over all, that eliminates the somewhat 'magic' behavior in my commit from yesterday where setReload/RestartRequired return the current stamp when was == STARTING. Instead any caller gets a unique stamp and there's no fragility where someday a maintainer breaks the current magic.

**Pull request merged**
